src/gtfs-rt/alerts.js

```diff
@@ -0,0 +1,144 @@
+import { DateTime } from 'luxon';
+import logger from '../lib/logger.js';
+
+// Merges relevant GTFS service alerts into a GraphHopper response.
+// Mutates the routeResult param.
+export function mergeAlertsIntoRoutes(alerts, routeResult) {
+  if (!routeResult || !routeResult.paths || !alerts) return routeResult;
+
+  // Alerts can be agency-wide but only for a specific route type operated by that agency,
+  // for example all cable cars operated by SFMTA but not buses or light rail.
+  const routeTypesByAgencyId = new Map();
+  // I'm assuming route, trip and stop IDs are globally unique, without having to
+  // join them with the agency ID.
+  const routeIds = new Set();
+  const tripIds = new Set();
+  const boardAndAlightStopIds = new Set();
+
+  let earliestDeparture;
+  let latestArrival;
+
+  for (const path of routeResult.paths) {
+    for (const leg of path.legs || []) {
+      if (leg.type !== 'pt') continue;
+
+      if (typeof leg.departure_time === 'string') {
+        const timestamp = DateTime.fromISO(leg.departure_time).toMillis();
+        if (!earliestDeparture || timestamp < earliestDeparture)
+          earliestDeparture = timestamp;
+      }
+
+      if (typeof leg.arrival_time === 'string') {
+        const timestamp = DateTime.fromISO(leg.arrival_time).toMillis();
+        if (!latestArrival || timestamp > latestArrival)
+          latestArrival = timestamp;
+      }
+
+      // add agency ID and route type combo
+      if (typeof leg.agency_id === 'string') {
+        if (!routeTypesByAgencyId.has(leg.agency_id))
+          routeTypesByAgencyId.set(leg.agency_id, new Set());
+        routeTypesByAgencyId.get(leg.agency_id).add(leg.route_type);
+      }
+
+      if (typeof leg.route_id === 'string')
+        routeIds.add(leg.route_id);
+
+      if (typeof leg.trip_id === 'string')
+        tripIds.add(leg.trip_id);
+
+      if (Array.isArray(leg.stops) && leg.stops.length > 0) {
+        const boardStopId = leg.stops[0]?.stop_id;
+        const alightStopId = leg.stops[leg.stops.length - 1]?.stop_id;
+        if (typeof boardStopId === 'string')
+          boardAndAlightStopIds.add(boardStopId);
+        if (typeof alightStopId === 'string')
+          boardAndAlightStopIds.add(alightStopId);
+      }
+    }
+  }
+
+  const relevantAlerts = alerts.filter(alert => {
```

Comment:
thanks for figuring out all these rules. 🙌 . personally I prefer a bunch of small functions then a chain of `.filter()`s that applies each of them separately. This mostly come from wanting to make unit testing cleaner/simpler. at some point I'll add tests to this repo at which point I will likely refactor this a bit. but we can discuss it then if I ever get around to it.